blog/modules/ROOT/pages/using-types.adoc

```diff
@@ -0,0 +1,220 @@
+
+[source,pkl]
+----
+serverPort: UInt16 // <1>
+----
+<1> Will only allow values between 0 and 65535.
+
+Builtin typealiases are not the only way to handle such cases, though.
+Numbers can be constrained to any value.
+
+[source,pkl]
+----
+volume: Float(isBetween(0.0, 11.0)) // <1>
+
+multiplier: Int(isEven) // <2>
+----
+<1> `volume` can only be in the range between 0 and 11.
+<2> `multiplier` cannot be odd.
+
+== Beyond builtins
+
+Up to now, we only took a look at builtin functions, mostly functions of the type itself.
+We also limited our types to a single constraint. Pkl allows, though, for arbitrary functions as constraints,
+and we can provide multiple ones:
+
+[source,pkl]
+----
+local function isPrime(n: Int): Boolean = !IntSeq(2, n - 1).toList().any((x) -> n % x == 0)
+
+hash: Int(isPrime(this)) // <1>
+
+name: String(this == capitalize()) // <2>
+
+building: String(isEmpty || this.toInt() <= 100) // <2>
+
+region: String(startsWith("us-"), length < 10) // <3>
+
+grade: Int(isBetween(0, 10))|String(matches(Regex(#"[A|B|C|D|E|F][+-]?"#))) // <4>
+----
+<1> User defined constraints.
+<2> Arbitrary boolean expressions.
+<3> Multiple constraints.
+<4> Multiple possibilities with different constraints.
+
+== Better errors
+
+One downside of using arbitrary expressions as constraints is that, sometimes, errors messages
+can be a bit hard to understand:
+
+[source,pkl]
+----
+ssn: String(matches(Regex(#"\d{3}-\d{2}-\d{4}"#))) = "123-123-123"
+----
+
+[source,text]
+----
+–– Pkl Error ––
+Type constraint `matches(Regex(#"\d{3}-\d{2}-\d{4}"#))` violated.
+Value: "123-123-123"
+
+1 | ssn: String(matches(Regex(#"\d{3}-\d{2}-\d{4}"#))) = "123-123-123"
+                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+...
+----
+
+We can do better than that.
+
+One trick to improve error messages is to create named functions:
+
+[source,pkl]
+----
+local function isValidSocialSecurityNumber(str: String): Boolean =
+  str.matches(Regex(#"\d{3}-\d{2}-\d{4}"#))
+
+ssn: String(isValidSocialSecurityNumber(this)) = "123-123-123"
+----
+
+[source,text]
+----
+–– Pkl Error ––
+Type constraint `isValidSocialSecurityNumber(this)` violated.
+Value: "123-123-123"
+
+4 | ssn: String(isValidSocialSecurityNumber(this)) = "123-123-123"
+                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+...
+----
+
+That's already better. Users now have a hint of what they did wrong.
+
+=== Providing custom error messages
+
+We just saw how to improve error messages with functions, but Pkl can
+do even better than that.
+
+With the knowledge that failing constraints will throw an exception, we can provide a custom
+error message by throwing the exception ourselves:
+
+[source,pkl]
+----
+local function reportSSN(ssn: String) =
+  """
+  Invalid social security number: \(ssn).
+  Valid ones should be in the form `XXX-XX-XXXX`
+  where `X` is a number between 0 and 9.
+  """
+
+ssn: String(matches(Regex(#"\d{3}-\d{2}-\d{4}"#)) || throw(reportSSN(this))) = "123-123-123"
```

Comment:
The one downside of this approach is that it does not compose well in type unions (because exceptions are fatal), i.e.
```
typealias SSN = String(matches(Regex(#"\d{3}-\d{2}-\d{4}"#)) || throw(reportSSN(this)))
typealias NIN = String(matches(Regex(#"[A-Z]{2}\d{6}[A-Z]"#)) || throw(reportNIN(this))
typealias TaxIdUSorUK = SSN | NIN // <- actually doesn't work
```

It's absolutely valuable to share this pattern of `|| throw` here, but I think you should point out this limitation.

(I like the example, btw)

Reply:
Good point.

Reply:
This pattern is somewhat of a footgun because of this issue. It's useful, but kind of dangerous because as a template author it's easy to define a union type without realizing that it's not _actually_ a union.

Reply:
I added a section to explain the downsides of using `throw` inside constraints.